### ISSUE TITLE
3.0 breaks connections using ansible

### ISSUE THREAD

**Issue by alfredodeza** [philpep/testinfra#439]
I'm trying out version 3.0.1 with `--connection=ansible` and it breaks on `ansible_runner`. The same command works fine with 2.1.0:

```
pytest -vvv --connection=ansible --ansible-inventory=hosts test_remote.py 
collected 0 items / 1 errors                                                  

=================================== ERRORS ====================================
_______________________ ERROR collecting test_remote.py _______________________
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/_pytest/python.py:238: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/_pytest/python.py:414: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/testinfra/plugin.py:113: in pytest_generate_tests
    ansible_inventory=metafunc.config.option.ansible_inventory,
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/testinfra/host.py:143: in get_hosts
    for backend in testinfra.backend.get_backends(hosts, **kwargs):
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/testinfra/backend/__init__.py:91: in get_backends
    for name in klass.get_hosts(host, **kw):
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/testinfra/backend/ansible.py:59: in get_hosts
    return AnsibleRunner.get_runner(inventory).get_hosts(host)
../../../../.virtualenvs/ceph-volume/lib/python3.6/site-packages/testinfra/utils/ansible_runner.py:124: in get_hosts
    groupmatch = fnmatch.fnmatch(group, pattern)
../../../../.virtualenvs/ceph-volume/lib/python3.6/fnmatch.py:35: in fnmatch
    pat = os.path.normcase(pat)
../../../../.virtualenvs/ceph-volume/lib/python3.6/posixpath.py:54: in normcase
    s = os.fspath(s)
E   TypeError: expected str, bytes or os.PathLike object, not NoneType
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
```

With 2.1.0 is fine:

```
pytest -vvv --connection=ansible --ansible-inventory=hosts test_remote.py
============================= test session starts =============================
platform linux2 -- Python 2.7.16, pytest-4.4.1, py-1.8.0, pluggy-0.10.0 -- /tmp/27/bin/python2
cachedir: .pytest_cache
rootdir: /home/alfredo/python/
plugins: testinfra-2.1.0
collected 2 items                                                             

test_remote.py::test_release_file[ansible://node2] PASSED               [ 50%]
test_remote.py::test_release_file[ansible://node3] PASSED               [100%]
```

The test file:

```
def test_release_file(host):
    release_file = host.file("/etc/os-release")
    assert release_file.contains('CentOS')
    assert release_file.contains('VERSION="7 (Core)"')
```

And the hosts file:

```
[nginx]
node2
node3
```

**Comment by alfredodeza** [comment]
If I try passing `--hosts=all` testinfra assumes an SSH connection:

```
transport.py              1746 ERROR    Exception: key cannot be used for signing
transport.py              1744 ERROR    Traceback (most recent call last):
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/transport.py", line 2036, in run
transport.py              1744 ERROR        handler(self.auth_handler, m)
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/auth_handler.py", line 298, in _parse_service_accept
transport.py              1744 ERROR        sig = self.private_key.sign_ssh_data(blob)
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/agent.py", line 418, in sign_ssh_data
transport.py              1744 ERROR        raise SSHException("key cannot be used for signing")
transport.py              1744 ERROR    paramiko.ssh_exception.SSHException: key cannot be used for signing
transport.py              1744 ERROR    
transport.py              1746 ERROR    Exception: key cannot be used for signing
transport.py              1744 ERROR    Traceback (most recent call last):
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/transport.py", line 2036, in run
transport.py              1744 ERROR        handler(self.auth_handler, m)
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/auth_handler.py", line 298, in _parse_service_accept
transport.py              1744 ERROR        sig = self.private_key.sign_ssh_data(blob)
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/agent.py", line 418, in sign_ssh_data
transport.py              1744 ERROR        raise SSHException("key cannot be used for signing")
transport.py              1744 ERROR    paramiko.ssh_exception.SSHException: key cannot be used for signing
transport.py              1744 ERROR
```

**Comment by alfredodeza** [comment]
If I explicitly use `--hosts='ansible://node1` (node1 is a VM) it thinks it is a local connection (the test fails because the local machine is not CentOS):

```
pytest -vvv --hosts='ansible://node1' test_remote.py
============================= test session starts =============================
platform linux -- Python 3.6.8, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /home/alfredo/.virtualenvs/ceph-volume/bin/python3.6
cachedir: .pytest_cache
rootdir: /home/alfredo/python/
plugins: testinfra-3.0.0, xdist-1.28.0, forked-1.0.2
collected 1 item                                                              

test_remote.py::test_release_file[ansible://localhost] FAILED           [100%]

================================== FAILURES ===================================
___________________ test_release_file[ansible://localhost] ____________________

host = <testinfra.host.Host object at 0x7fc4beeb7400>

    def test_release_file(host):
        release_file = host.file("/etc/os-release")
>       assert release_file.contains('CentOS')
E       AssertionError: assert False
E        +  where False = <bound method File.contains of <file /etc/os-release>>('CentOS')
E        +    where <bound method File.contains of <file /etc/os-release>> = <file /etc/os-release>.contains

test_remote.py:4: AssertionError
------------------------------ Captured log call ------------------------------
base.py                    268 INFO     RUN CommandResult(command=b'uname -s', exit_status=0, stdout=b'Linux\n', stderr=None)
base.py                    268 INFO     RUN CommandResult(command=b'lsb_release -a', exit_status=0, stdout=b'Distributor ID:\tUbuntu\nDescription:\tUbuntu 19.04\nRelease:\t19.04\nCodename:\tdisco\n', stderr=b'No LSB modules are available.\n')
base.py                    268 INFO     RUN CommandResult(command=b'grep -qs -- CentOS /etc/os-release', exit_status=1, stdout=None, stderr=None)
========================== 1 failed in 0.79 seconds ===========================
```

**Comment by philpep** [comment]
Hi, from traceback code, it looks you're still using testinfra 3.0.0 and not 3.0.1

Can you retry with testinfra 3.0.1 ?

That being said, the behavior you get is very weird even with bogus 3.0.0, I cannot reproduce locally:

```
$ virtualenv -p python3 v
$ source v/bin/activate
$ pip install testinfra ansible
$ echo "$(hostname)" > inventory
$ cat << EOF > t.py
def test_dummy(host):
    assert host.run('false').rc == 1
EOF
# euler is my hostname
$ python -m pytest -v --connection=ansible --ansible-inventory=inventory t.py
t.py::test_dummy[ansible://euler] PASSED
$ python -m pytest -v --connection=ansible --ansible-inventory=inventory --hosts=all t.py
t.py::test_dummy[ansible://euler] PASSED
$ python -m pytest -v --ansible-inventory=inventory --hosts=ansible://all t.py
t.py::test_dummy[ansible://euler] PASSED
$ python -m pytest -v --ansible-inventory=inventory --hosts=ansible://euler t.py
t.py::test_dummy[ansible://euler] PASSED
$ python -m pytest -v '--hosts=ansible://euler?ansible_inventory=inventory' t.py
t.py::test_dummy[ansible://euler] PASSED
$ python -m pytest -v --hosts=ansible://localhost t.py
t.py::test_dummy[ansible://localhost] PASSED
$ python -m pytest -v --hosts=ansible://euler t.py
t.py::test_dummy[ansible://localhost] PASSED
```

Only the latest test case is wrong, because when inventory is empty (not specified), it return a local connection to "localhost" which doesn't match the given pattern. We need to fix this.

**Comment by alfredodeza** [comment]
I've upgraded to 3.0.1 with similar results :( 

The output from pytest is very long, but these two makes me assume that keys aren't being loaded properly:

```
test_remote.py::test_release_file[ansible://node3]
  /home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/client.py:822: UserWarning: Unknown ssh-ed25519 host key for node3: b'52cdb9e883c2548b6ea68fe292db0eeb'
    key.get_name(), hostname, hexlify(key.get_fingerprint())
---

And this one is the same as before with 3.0.0

```
transport.py              1746 ERROR    Exception: key cannot be used for signing
transport.py              1744 ERROR    Traceback (most recent call last):
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/transport.py", line 2036, in run
transport.py              1744 ERROR        handler(self.auth_handler, m)
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/auth_handler.py", line 298, in _parse_service_accept
transport.py              1744 ERROR        sig = self.private_key.sign_ssh_data(blob)
transport.py              1744 ERROR      File "/home/alfredo/.virtualenvs/ceph-volume/lib/python3.6/site-packages/paramiko/agent.py", line 418, in sign_ssh_data
transport.py              1744 ERROR        raise SSHException("key cannot be used for signing")
transport.py              1744 ERROR    paramiko.ssh_exception.SSHException: key cannot be used for signing
transport.py              1744 ERROR
```

**Comment by philpep** [comment]
Hmm, where do you define ssh keys ? In inventory, ansible.cfg, environment variables ?

How can I reproduce the issue ? Any link to a test suite on github or travis/jenkins build ?

**Comment by alfredodeza** [comment]
My SSH keys are in `$HOME/.ssh/config` and 2.1.0 just works without specifying anything. I will try to get you a reproducer - these are all reproducible locally with new virtualenvs.

I will try with a vanilla VM and see if I can reproduce. Not sure what is special about my machine

**Comment by alfredodeza** [comment]
Ok I have a bit of a reproducer for at least one of the problems. CentOS 7, Python 2.7:

```
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

```

Here are the versions installed in the virtualenv:

```
ansible (2.7.10)
asn1crypto (0.24.0)
atomicwrites (1.3.0)
attrs (19.1.0)
bcrypt (3.1.6)
cffi (1.12.3)
cryptography (2.6.1)
enum34 (1.1.6)
funcsigs (1.0.2)
ipaddress (1.0.22)
Jinja2 (2.10.1)
MarkupSafe (1.1.1)
more-itertools (5.0.0)
paramiko (2.4.2)
pathlib2 (2.3.3)
pip (9.0.1)
pluggy (0.11.0)
py (1.8.0)
pyasn1 (0.4.5)
pycparser (2.19)
PyNaCl (1.3.0)
pytest (4.4.1)
PyYAML (5.1)
scandir (1.10.0)
setuptools (28.8.0)
six (1.12.0)
testinfra (3.0.1)
wheel (0.29.0)
```

The test:

```
$ cat test_remote.py 
   
def test_release_file(host):
    release_file = host.file("/etc/os-release")
    assert release_file.contains('CentOS')
    assert release_file.contains('VERSION="7 (Core)"') 
```

The invocation shows I am specifying an IP which is *remote* (not the local one), and yet it connects to localhost:

```
$ pytest -v --hosts='ansible://192.168.111.102' test_remote.py
============================= test session starts =============================
platform linux2 -- Python 2.7.5, pytest-4.4.1, py-1.8.0, pluggy-0.11.0 -- /home/vagrant/tmp/bin/python
cachedir: .pytest_cache
rootdir: /home/vagrant/tmp
plugins: testinfra-3.0.1
collected 1 item                                                              

test_remote.py::test_release_file[ansible://localhost] PASSED           [100%]

```

**Comment by philpep** [comment]
Yes, you have to pass an ansible inventory here (with ``--ansible-inventory`` or by adding ``?ansible_inventory=...`` to the ``--host`` option) or testinfra will use the default from ``ansible-inventory --list``.

I'll fix this behavior to have the same behavior than testinfra 2.X which skip the test.

**Comment by alfredodeza** [comment]
Ok. Same vanilla CentOS server, trying to connect to "node3" which lives at 192.168.111.102:

```
(tmp) [vagrant@node2 tmp]$ cat hosts
[all]
node3
#192.168.111.102
(tmp) [vagrant@node2 tmp]$ cat ~/.ssh/config
Host node3
  HostName 192.168.111.102
  User vagrant
  Port 22
  UserKnownHostsFile /dev/null
  StrictHostKeyChecking no
  PasswordAuthentication no
  IdentityFile /home/vagrant/insecure_private_key
  IdentitiesOnly yes
#  LogLevel FATAL
(tmp) [vagrant@node2 tmp]$ pytest --hosts=all --connection=ansible --ansible-inventory=hosts test_remote.py
=============================================================================================== test session starts ===============================================================================================
platform linux2 -- Python 2.7.5, pytest-4.4.1, py-1.8.0, pluggy-0.11.0
rootdir: /home/vagrant/tmp
plugins: testinfra-3.0.1
collected 1 item                                                                                                                                                                                                  

test_remote.py F                                                                                                                                                                                            [100%]

==================================================================================================== FAILURES =====================================================================================================
_______________________________________________________________________________________ test_release_file[ansible://node3] ________________________________________________________________________________________

host = <testinfra.host.Host object at 0x7fb92bcfc610>

    def test_release_file(host):
>       release_file = host.file("/etc/os-release")

test_remote.py:3: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lib/python2.7/site-packages/testinfra/host.py:107: in __getattr__
    obj = module_class.get_module(self)
lib/python2.7/site-packages/testinfra/modules/base.py:22: in get_module
    klass = cls.get_module_class(_host)
lib/python2.7/site-packages/testinfra/modules/file.py:187: in get_module_class
    if host.system_info.type == "linux":
lib/python2.7/site-packages/testinfra/modules/systeminfo.py:143: in type
    return self.sysinfo["type"]
lib/python2.7/site-packages/testinfra/utils/__init__.py:44: in __get__
    value = obj.__dict__[self.func.__name__] = self.func(obj)
lib/python2.7/site-packages/testinfra/modules/systeminfo.py:33: in sysinfo
    uname = self.run_expect([0, 1], 'uname -s')
lib/python2.7/site-packages/testinfra/host.py:71: in run
    return self.backend.run(command, *args, **kwargs)
lib/python2.7/site-packages/testinfra/backend/ansible.py:41: in run
    return self.ansible_runner.run(self.host, command)
lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:167: in run
    return self.get_host(host).run(command)
lib/python2.7/site-packages/testinfra/host.py:71: in run
    return self.backend.run(command, *args, **kwargs)
lib/python2.7/site-packages/testinfra/backend/paramiko.py:103: in run
    if not self.client.get_transport().is_active():
lib/python2.7/site-packages/testinfra/utils/__init__.py:44: in __get__
    value = obj.__dict__[self.func.__name__] = self.func(obj)
lib/python2.7/site-packages/testinfra/backend/paramiko.py:84: in client
    client.connect(**cfg)
lib/python2.7/site-packages/paramiko/client.py:437: in connect
    passphrase,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <paramiko.client.SSHClient object at 0x7fb928ea7a50>, username = 'vagrant', password = None, pkey = None, key_filenames = [], allow_agent = True, look_for_keys = True, gss_auth = False, gss_kex = False
gss_deleg_creds = True, gss_host = None, passphrase = None

    def _auth(
        self,
        username,
        password,
        pkey,
        key_filenames,
        allow_agent,
        look_for_keys,
        gss_auth,
        gss_kex,
        gss_deleg_creds,
        gss_host,
        passphrase,
    ):
        """
        Try, in order:
    
            - The key(s) passed in, if one was passed in.
            - Any key we can find through an SSH agent (if allowed).
            - Any "id_rsa", "id_dsa" or "id_ecdsa" key discoverable in ~/.ssh/
              (if allowed).
            - Plain username/password auth, if a password was given.
    
        (The password might be needed to unlock a private key [if 'passphrase'
        isn't also given], or for two-factor authentication [for which it is
        required].)
        """
        saved_exception = None
        two_factor = False
        allowed_types = set()
        two_factor_types = {"keyboard-interactive", "password"}
        if passphrase is None and password is not None:
            passphrase = password
    
        # If GSS-API support and GSS-PI Key Exchange was performed, we attempt
        # authentication with gssapi-keyex.
        if gss_kex and self._transport.gss_kex_used:
            try:
                self._transport.auth_gssapi_keyex(username)
                return
            except Exception as e:
                saved_exception = e
    
        # Try GSS-API authentication (gssapi-with-mic) only if GSS-API Key
        # Exchange is not performed, because if we use GSS-API for the key
        # exchange, there is already a fully established GSS-API context, so
        # why should we do that again?
        if gss_auth:
            try:
                return self._transport.auth_gssapi_with_mic(
                    username, gss_host, gss_deleg_creds
                )
            except Exception as e:
                saved_exception = e
    
        if pkey is not None:
            try:
                self._log(
                    DEBUG,
                    "Trying SSH key {}".format(
                        hexlify(pkey.get_fingerprint())
                    ),
                )
                allowed_types = set(
                    self._transport.auth_publickey(username, pkey)
                )
                two_factor = allowed_types & two_factor_types
                if not two_factor:
                    return
            except SSHException as e:
                saved_exception = e
    
        if not two_factor:
            for key_filename in key_filenames:
                for pkey_class in (RSAKey, DSSKey, ECDSAKey, Ed25519Key):
                    try:
                        key = self._key_from_filepath(
                            key_filename, pkey_class, passphrase
                        )
                        allowed_types = set(
                            self._transport.auth_publickey(username, key)
                        )
                        two_factor = allowed_types & two_factor_types
                        if not two_factor:
                            return
                        break
                    except SSHException as e:
                        saved_exception = e
    
        if not two_factor and allow_agent:
            if self._agent is None:
                self._agent = Agent()
    
            for key in self._agent.get_keys():
                try:
                    id_ = hexlify(key.get_fingerprint())
                    self._log(DEBUG, "Trying SSH agent key {}".format(id_))
                    # for 2-factor auth a successfully auth'd key password
                    # will return an allowed 2fac auth method
                    allowed_types = set(
                        self._transport.auth_publickey(username, key)
                    )
                    two_factor = allowed_types & two_factor_types
                    if not two_factor:
                        return
                    break
                except SSHException as e:
                    saved_exception = e
    
        if not two_factor:
            keyfiles = []
    
            for keytype, name in [
                (RSAKey, "rsa"),
                (DSSKey, "dsa"),
                (ECDSAKey, "ecdsa"),
                (Ed25519Key, "ed25519"),
            ]:
                # ~/ssh/ is for windows
                for directory in [".ssh", "ssh"]:
                    full_path = os.path.expanduser(
                        "~/{}/id_{}".format(directory, name)
                    )
                    if os.path.isfile(full_path):
                        # TODO: only do this append if below did not run
                        keyfiles.append((keytype, full_path))
                        if os.path.isfile(full_path + "-cert.pub"):
                            keyfiles.append((keytype, full_path + "-cert.pub"))
    
            if not look_for_keys:
                keyfiles = []
    
            for pkey_class, filename in keyfiles:
                try:
                    key = self._key_from_filepath(
                        filename, pkey_class, passphrase
                    )
                    # for 2-factor auth a successfully auth'd key will result
                    # in ['password']
                    allowed_types = set(
                        self._transport.auth_publickey(username, key)
                    )
                    two_factor = allowed_types & two_factor_types
                    if not two_factor:
                        return
                    break
                except (SSHException, IOError) as e:
                    saved_exception = e
    
        if password is not None:
            try:
                self._transport.auth_password(username, password)
                return
            except SSHException as e:
                saved_exception = e
        elif two_factor:
            try:
                self._transport.auth_interactive_dumb(username)
                return
            except SSHException as e:
                saved_exception = e
    
        # if we got an auth-failed exception earlier, re-raise it
        if saved_exception is not None:
            raise saved_exception
>       raise SSHException("No authentication methods available")
E       SSHException: No authentication methods available

lib/python2.7/site-packages/paramiko/client.py:750: SSHException
================================================================================================ warnings summary =================================================================================================
test_remote.py::test_release_file[ansible://node3]
  /home/vagrant/tmp/lib/python2.7/site-packages/paramiko/kex_ecdh_nist.py:39: CryptographyDeprecationWarning: encode_point has been deprecated on EllipticCurvePublicNumbers and will be removed in a future version. Please use EllipticCurvePublicKey.public_bytes to obtain both compressed and uncompressed point encoding.
    m.add_string(self.Q_C.public_numbers().encode_point())

test_remote.py::test_release_file[ansible://node3]
  /home/vagrant/tmp/lib/python2.7/site-packages/paramiko/kex_ecdh_nist.py:96: CryptographyDeprecationWarning: Support for unsafe construction of public numbers from encoded data will be removed in a future version. Please use EllipticCurvePublicKey.from_encoded_point
    self.curve, Q_S_bytes

test_remote.py::test_release_file[ansible://node3]
  /home/vagrant/tmp/lib/python2.7/site-packages/paramiko/kex_ecdh_nist.py:111: CryptographyDeprecationWarning: encode_point has been deprecated on EllipticCurvePublicNumbers and will be removed in a future version. Please use EllipticCurvePublicKey.public_bytes to obtain both compressed and uncompressed point encoding.
    hm.add_string(self.Q_C.public_numbers().encode_point())

test_remote.py::test_release_file[ansible://node3]
  /home/vagrant/tmp/lib/python2.7/site-packages/paramiko/client.py:822: UserWarning: Unknown ssh-ed25519 host key for node3: 52cdb9e883c2548b6ea68fe292db0eeb
    key.get_name(), hostname, hexlify(key.get_fingerprint())

-- Docs: https://docs.pytest.org/en/latest/warnings.html
====================================================================================== 1 failed, 4 warnings in 1.05 seconds =======================================================================================
(tmp) [vagrant@node2 tmp]$ ssh node3
Warning: Permanently added '192.168.111.102' (ECDSA) to the list of known hosts.
Last login: Tue May  7 19:36:02 2019 from 192.168.111.101
[vagrant@node3 ~]$ logout
Connection to 192.168.111.102 closed.
```

So I can ssh without problems to node3 (192.168.111.102) but can't with pytest/testinfra even with --ansible-inventory. I tried both the IP and the hostname with the same behavior.

Would that setup be enough for a reproducer? Have you tried a fresh/vanilla CentOS VM?

**Comment by philpep** [comment]
Oh I see, this is because testinfra with paramiko backend doesn't read ~/.ssh/config, does it works if you change "paramiko" by "ssh" here https://github.com/philpep/testinfra/blob/master/testinfra/utils/ansible_runner.py#L81 (or just drop the "if" statement).

I think reading user ~/.ssh/config by default in paramiko connection seems more reasonable than using plain ssh calls, I'll make this change.

**Comment by alfredodeza** [comment]
if I change that line to `'ssh'` it works *like a champ*!

I think reading ~/.ssh/config makes sense, but I would also like to add that using `--ssh-config=~/.ssh/config` didn't had any effect

Thank you so much for your patience while trying to fix this!

**Comment by philpep** [comment]
Yes I make sense to support ``--ssh-config`` and ``--ssh-identity-file`` here as well. I'll make this.
Thanks for reporting this bug, I this will save time to a lot of users using testinfra+ansible !

**Comment by philpep** [comment]
I just released 3.0.2 with these fixes, tell me if it works for you.

**Comment by alfredodeza** [comment]
3.0.2 works! Thank you so much
